cdc: Retry on certain sink failures

mrtracy opened the pull request:
Allow changefeeds to retry when emitting to a sink fails under certain
circumstances. The following circumstances are currently considered
retryable:

+ Any ProducerError from Sarama when emitting to Kafka. This may cause
us to retry in some cases that may never succeed, but we don't yet have
enough information do identify these situations, so we take the liberal
position and always retry.
+ Using the experimental sql sink in combination with a new testhook for
throwing fake retryable errors.

Implementation works by detecting retryable errors in
changefeedResumer.Resume(), and re-establishing a new distsql Changefeed
flow using the updated parameters.

Includes a new metric which counts the total number of retry attempts
for sink errors.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/29563)
<!-- Reviewable:end -->

danhhz reviewed (commented):
This is looking great! I'm a little bit in shock that your test didn't catch any bugs in the impl

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29563#-:-LLbHReHAnZmiH1VpkEh:bj588b2)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 120 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29563#-LLbGOfxDQzJZtLVv9A2:-LLbGOfxDQzJZtLVv9A3:b-qhurqy) ([raw file](https://github.com/cockroachdb/cockroach/blob/d969cc7663d9d76a255e318a4a54acc1cd932d11/pkg/ccl/changefeedccl/changefeed_processors.go#L120)):*
> ```Go
> 		ca.changedRowBuf = &st.buf
> 	case *sqlSink:
> 		st.failHook = flowCtx.TestingKnobs().FailChangefeedSqlSinkHook
> ```

if you do my `AfterSinkFlush` suggestion below, we can avoid adding another special case
___
*[pkg/ccl/changefeedccl/changefeed_stmt.go, line 363 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29563#-LLbGEOXAxkYWiN5JQEl:-LLbGEOXAxkYWiN5JQEm:bh6rsrc) ([raw file](https://github.com/cockroachdb/cockroach/blob/d969cc7663d9d76a255e318a4a54acc1cd932d11/pkg/ccl/changefeedccl/changefeed_stmt.go#L363)):*
> ```Go
> 		// a dummy channel.
> 		startedCh = make(chan tree.Datums, 1)
> 		log.Infof(ctx, `CHANGEFEED job %d encountered retryable error: %v`, *job.ID(), err)
> ```

nit: move this line right below the `if !isRetryableSinkError(err) {` block so it gets logged even if there's an issue reloading the job info
___
*[pkg/ccl/changefeedccl/changefeed_stmt.go, line 387 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29563#-LLbFhPW6PIaqis2UObA:-LLbFhPW6PIaqis2UObB:bs2es5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d969cc7663d9d76a255e318a4a54acc1cd932d11/pkg/ccl/changefeedccl/changefeed_stmt.go#L387)):*
> ```Go
> 	// possible Kafka errors to make this determination, so for now we default to
> 	// treating all such errors as retryable.
> 	if _, ok := err.(*sarama.ProducerError); ok {
> ```

This feels like too high of a level for sarama knowledge. How about something like the pkg/errors pattern

```
type causer interface {
       Cause() error
}

type retryableSinkError struct {
       cause error
}

func (e retryableSinkError) Error() string { return e.cause.Error() }
func (e retryableSinkError) Cause() error  { return e.cause }

func isRetryablySinkError(err error) bool {
       for {
               if _, ok := err.(retryableSinkError); ok {
                       return true
               }
               if e, ok := err.(causer); ok {
                       err = e.Cause()
                       continue
               }
               return false
       }
}
```

Then the sink itself is responsible for wrapping its retryable errors (or all of them) in one of these before it returns. This composes well with my AfterSinkFlush suggestion
___
*[pkg/ccl/changefeedccl/metrics.go, line 121 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29563#-LLbGXnz9dsVjSQqWyqj:-LLbGXnz9dsVjSQqWyqk:b-d7up1k) ([raw file](https://github.com/cockroachdb/cockroach/blob/d969cc7663d9d76a255e318a4a54acc1cd932d11/pkg/ccl/changefeedccl/metrics.go#L121)):*
> ```Go
> 	}
> 	metaChangefeedSinkErrorRetries = metric.Metadata{
> 		Name:        "changefeed.sink_error_retries",
> ```

awesome! I think this is a good metric to have anyway
___
*[pkg/sql/distsqlrun/server.go, line 613 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29563#-LLbGey16Q8Vf0_ErF5R:-LLbGey16Q8Vf0_ErF5S:b-7s18v0) ([raw file](https://github.com/cockroachdb/cockroach/blob/d969cc7663d9d76a255e318a4a54acc1cd932d11/pkg/sql/distsqlrun/server.go#L613)):*
> ```Go
> 	DeterministicStats bool
> 
> 	// FailChangefeedSqlSink is a function that can be used to cause the SQL
> ```

Instead of adding our cruft to the distsql hooks struct, let's add a single `ChangefeedKnobs base.ModuleTestingKnobs` 
field that in practice is populated with a new `changefeedccl.TestingKnobs` struct. (I know, I know. You'll have to type assert it inside the changefeedccl package, but it's definitely cleaner this way.)
___
*[pkg/sql/distsqlrun/server.go, line 616 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29563#-LLbEJChECrUUxDVHsk7:-LLbEJChECrUUxDVHsk8:b-4juij) ([raw file](https://github.com/cockroachdb/cockroach/blob/d969cc7663d9d76a255e318a4a54acc1cd932d11/pkg/sql/distsqlrun/server.go#L616)):*
> ```Go
> 	// Changefeed sink to fail If this function is set and returns true, emits to
> 	// SQL Sinks will fail with a retryable error.
> 	FailChangefeedSqlSinkHook func() bool
> ```

In the next day or two, I'm going to need a hook that runs after Flush (so I can get notifications for when the sql sink has new data). We could pretty easily combine the two and have a hook that runs right after (the single callsite of) flush and returns an error.

At https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/changefeedccl/changefeed.go#L235, something like
```
if err := sink.Flush(ctx); err != nil {
  return nil, err
}
if err := knobs.AfterSinkFlush(); err != nil {
  return nil, err
}
```


<!-- Sent from Reviewable.io -->

mrtracy reviewed (commented):
RFAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29563#-:-LLfTcfG9e5sCh258ybU:bb72pd0)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 120 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29563#-LLbGOfxDQzJZtLVv9A2:-LLfPXaG3f-iCUCy0mBO:bjx7f9r) ([raw file](https://github.com/cockroachdb/cockroach/blob/d969cc7663d9d76a255e318a4a54acc1cd932d11/pkg/ccl/changefeedccl/changefeed_processors.go#L120)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

if you do my `AfterSinkFlush` suggestion below, we can avoid adding another special case
</blockquote></details>

Done as part of the AfterSinkFlush work suggested.
___
*[pkg/ccl/changefeedccl/changefeed_stmt.go, line 363 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29563#-LLbGEOXAxkYWiN5JQEl:-LLebGkH0Y9KHjGNI75H:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/d969cc7663d9d76a255e318a4a54acc1cd932d11/pkg/ccl/changefeedccl/changefeed_stmt.go#L363)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: move this line right below the `if !isRetryableSinkError(err) {` block so it gets logged even if there's an issue reloading the job info
</blockquote></details>

Done
___
*[pkg/ccl/changefeedccl/changefeed_stmt.go, line 387 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29563#-LLbFhPW6PIaqis2UObA:-LLfTQ5kAy-vlICmtE9l:bok5uap) ([raw file](https://github.com/cockroachdb/cockroach/blob/d969cc7663d9d76a255e318a4a54acc1cd932d11/pkg/ccl/changefeedccl/changefeed_stmt.go#L387)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

This feels like too high of a level for sarama knowledge. How about something like the pkg/errors pattern

```
type causer interface {
       Cause() error
}

type retryableSinkError struct {
       cause error
}

func (e retryableSinkError) Error() string { return e.cause.Error() }
func (e retryableSinkError) Cause() error  { return e.cause }

func isRetryablySinkError(err error) bool {
       for {
               if _, ok := err.(retryableSinkError); ok {
                       return true
               }
               if e, ok := err.(causer); ok {
                       err = e.Cause()
                       continue
               }
               return false
       }
}
```

Then the sink itself is responsible for wrapping its retryable errors (or all of them) in one of these before it returns. This composes well with my AfterSinkFlush suggestion
</blockquote></details>

Implemented this as given, moved it into "sink.go".  kafkaSink now wraps ProducerErrors inside of this error; I think it's appropriate for the sink itself to know about sarama.
___
*[pkg/sql/distsqlrun/server.go, line 613 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29563#-LLbGey16Q8Vf0_ErF5R:-LLel1fz0WcH0VmKtO-q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d969cc7663d9d76a255e318a4a54acc1cd932d11/pkg/sql/distsqlrun/server.go#L613)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Instead of adding our cruft to the distsql hooks struct, let's add a single `ChangefeedKnobs base.ModuleTestingKnobs` 
field that in practice is populated with a new `changefeedccl.TestingKnobs` struct. (I know, I know. You'll have to type assert it inside the changefeedccl package, but it's definitely cleaner this way.)
</blockquote></details>

Done.
___
*[pkg/sql/distsqlrun/server.go, line 616 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29563#-LLbEJChECrUUxDVHsk7:-LLfP5TA5i7_XV2TEmcP:b6oz0u4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d969cc7663d9d76a255e318a4a54acc1cd932d11/pkg/sql/distsqlrun/server.go#L616)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

In the next day or two, I'm going to need a hook that runs after Flush (so I can get notifications for when the sql sink has new data). We could pretty easily combine the two and have a hook that runs right after (the single callsite of) flush and returns an error.

At https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/changefeedccl/changefeed.go#L235, something like
```
if err := sink.Flush(ctx); err != nil {
  return nil, err
}
if err := knobs.AfterSinkFlush(); err != nil {
  return nil, err
}
```

</blockquote></details>

Done, changed the FailSqlHook into this AfterSinkFlush hook and made the appropriate changes.


<!-- Sent from Reviewable.io -->

danhhz reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29563#-:-LLfWHeK0b_7vKxn5P_I:bg7igkv)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/ccl/changefeedccl/changefeed_stmt.go, line 387 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29563#-LLbFhPW6PIaqis2UObA:-LLfVqXA0m_MEMazor-C:bg9jwws) ([raw file](https://github.com/cockroachdb/cockroach/blob/d969cc7663d9d76a255e318a4a54acc1cd932d11/pkg/ccl/changefeedccl/changefeed_stmt.go#L387)):*
> I think it's appropriate for the sink itself to know about sarama.

agreed
___
*[pkg/ccl/changefeedccl/metrics.go, line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29563#-LLfVaGh3J1h7Er0iHwX:-LLfVaGiBDpXgjxBLyuE:b-k7h7lw) ([raw file](https://github.com/cockroachdb/cockroach/blob/d969cc7663d9d76a255e318a4a54acc1cd932d11/pkg/ccl/changefeedccl/metrics.go#L42)):*
> ```Go
> 		s.metrics.EmittedBytes.Inc(int64(len(key) + len(value)))
> 		s.metrics.EmitNanos.Inc(timeutil.Since(start).Nanoseconds())
> 	} else if isRetryableSinkError(err) {
> ```

you could leave this here if you wanted. i think it still would work? /shrug


<!-- Sent from Reviewable.io -->

mrtracy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29563)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/ccl/changefeedccl/metrics.go, line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29563#-LLfVaGh3J1h7Er0iHwX:-LLfbWGHA9f4RCmAuWzj:b-8w1i7u) ([raw file](https://github.com/cockroachdb/cockroach/blob/d969cc7663d9d76a255e318a4a54acc1cd932d11/pkg/ccl/changefeedccl/metrics.go#L42)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

you could leave this here if you wanted. i think it still would work? /shrug
</blockquote></details>

Unfortunately no, the fake error in tests isn't getting thrown from inside Flush anymore, so metricsSink doesn't see it.


<!-- Sent from Reviewable.io -->

danhhz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29563)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/ccl/changefeedccl/metrics.go, line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29563#-LLfVaGh3J1h7Er0iHwX:-LLjtFUO30M8Y9xFa7QM:br4284v) ([raw file](https://github.com/cockroachdb/cockroach/blob/d969cc7663d9d76a255e318a4a54acc1cd932d11/pkg/ccl/changefeedccl/metrics.go#L42)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Unfortunately no, the fake error in tests isn't getting thrown from inside Flush anymore, so metricsSink doesn't see it.
</blockquote></details>

ah right. ack


<!-- Sent from Reviewable.io -->

mrtracy reviewed (commented):
Some context for those following along - the tests for this change have exposed that there is a possibility to leak a goroutine from the sqlSink, tripping our leaktest check.  I am investigating how to plug this leak.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29563#-:-LLkVq5G40NIpCnw6mqM:bpkutx9)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

mrtracy commented:
Okay, I think I finally got the leaks under control.  Learned a lot about context cancellation in the process.

I have moved much of the code that was present in newChangeAggregatorProcessor() into ChangeAggregatorProcess.Start() - this ensures that sinks are not created in a way that they won't eventually be closed (the Close() method will not be called on the processor if the processor never starts, which can happen if an error occurs during setup).

@danhhz RFAL.

mrtracy commented:
In fact, I have not solved this issue.  Back at it then, I guess.

mrtracy commented:
I had to apply the same change to the ChangeFrontier processor as I had already applied to ChangeAggregator processor (moving sink creation into the Start method, so that the close() method will be called).

danhhz reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29563#-:-LM3ktXYESHeS3uoPJbD:bnfp4nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 110 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/29563#-LM3kRgeDo7xMtVwDVPa:-LM3kRgeDo7xMtVwDVPb:bh3tvqe) ([raw file](https://github.com/cockroachdb/cockroach/blob/c96db8a6a59c69141e7168a64d188371d0bc5585/pkg/ccl/changefeedccl/changefeed_processors.go#L110)):*
> ```Go
> func (ca *changeAggregator) Start(ctx context.Context) context.Context {
> 	ctx, ca.cancel = context.WithCancel(ctx)
> 	ctx = ca.StartInternal(ctx, changeAggregatorProcName)
> ```

you convinced me offline why this is here (and why it's safe), but i feel like most of the other processors have it at the bottom. comment would be nice, here and in cf
___
*[pkg/ccl/changefeedccl/sink.go, line 367 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/29563#-LM3knZl1J8j6eUhnVwg:-LM3knZl1J8j6eUhnVwh:b-l6mwxj) ([raw file](https://github.com/cockroachdb/cockroach/blob/c96db8a6a59c69141e7168a64d188371d0bc5585/pkg/ccl/changefeedccl/sink.go#L367)):*
> ```Go
> 	// partitions to something small but greater than 1.
> 	sqlSinkNumPartitions = 3
> 	// testRetryableSinkError is a constant error value which indicates a synthetic,
> ```

remove


<!-- Sent from Reviewable.io -->

mrtracy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29563)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 110 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/29563#-LM3kRgeDo7xMtVwDVPa:-LM3n8_v1KOYTJusitWM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c96db8a6a59c69141e7168a64d188371d0bc5585/pkg/ccl/changefeedccl/changefeed_processors.go#L110)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

you convinced me offline why this is here (and why it's safe), but i feel like most of the other processors have it at the bottom. comment would be nice, here and in cf
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/sink.go, line 367 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/29563#-LM3knZl1J8j6eUhnVwg:-LM3nC5ZCrMf11n9tkF_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c96db8a6a59c69141e7168a64d188371d0bc5585/pkg/ccl/changefeedccl/sink.go#L367)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

remove
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mrtracy commented:
Well, I squashed all of the leaked goroutines; now a completely different test is failing that wasn't failing before. One of the CDC bank acceptance tests is failing with fingerprint mismatches; @danhhz does that failure look familiar to you at all?

danhhz commented:
That's a check for the overall correctness of our output. It's hooked up to the validator. Not sure why it wasn't failing before, but maybe it's flaky or maybe it's now broken. You can stress it locally if you comment out https://github.com/cockroachdb/cockroach/blob/master/pkg/acceptance/test_acceptance.go#L51 and run something like
```
make stress PKG=./pkg/ccl/acceptanceccl TAGS=acceptance TESTS=CDC//Bank TESTFLAGS='-v -l $GOPATH/src/github.com/cockroachdb/cockroach/artifacts/acceptance -show-logs' TESTTIMEOUT=60s STRESSFLAGS="-p 2"
```

Commit:
cdc: Fix poller race condition

Fixes a potential race condition in the poller - it was possible for
the initial state of a table watched by a changefeed to be incorrectly
reported if the table was modified in between the CREATE CHANGEFEED
statement and the time when the poller started internally.

This is fixed by choosing the first "nextHighwater" for the poller at
the time when the CREATE CHANGEFEED statement was executed, rather than
the time when the poller starts asynchronously.

Release note: None

Commit:
cdc: Fix leaked goroutine from Sink

Fixes a possibility of a leaked goroutine when creating either a
ChangeAggregator or ChangeFrontier processor. This leak occurred because
the Sink was being creating in the `new` methods for these processors,
but the associated cleanup function which is responsible for Closing the
sink is only called if the processor is actually started. Therefore, in
the situation where an error occurred during flow setup, Sinks could be
dropped on the floor and left unclosed, leaking goroutines.

Fix is to move sink creation into the Start() method of changeAggregator
and changeFrontier, ensuring that sinks don't actually get created unless
it is guaranteed they will be closed.

Note that we still create a "Canary sink" in the `new` methods
exclusively in order to test for deterministic errors, such as
improperly specified options. This is necessary in order to return such
errors to the user when they issue a CREATE CHANGEFEED statement. The
canary sinks are immediately closed.

Release note: None

Commit:
cdc: Retry on certain sink failures

Allow changefeeds to retry when emitting to a sink fails under certain
circumstances. The following circumstances are currently considered
retryable:

+ Any ProducerError from Sarama when emitting to Kafka. This may cause
us to retry in some cases that may never succeed, but we don't yet have
enough information do identify these situations, so we take the liberal
position and always retry.
+ Using a new "AfterFlush()" test hook to throw a fake retryable error.

Implementation works by detecting retryable errors in
changefeedResumer.Resume(), and re-establishing a new distsql Changefeed
flow using the updated parameters.

Includes a new metric which counts the total number of retry attempts
for sink errors.

Release note (sql change): Changefeeds will retry, rather than abort, in
certain cases when failing to emit to a sink.

mrtracy commented:
Okay, finally tracked down the source of that last bug: a race condition involving the poller.  I have broken this commit into three commits, the first two being general fixes which were exposed by the test in the actual Retry feature commit.

@danhhz I know you've already LGTMed this, but you probably want to take a look at the poller fix.

danhhz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29563)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/ccl/changefeedccl/poller.go, line 63 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/29563#-LMDH54c32sKa8NMsrhI:-LMDH54c32sKa8NMsrhJ:b-50a1il) ([raw file](https://github.com/cockroachdb/cockroach/blob/12f642e9eb643008fbd0ac7e25b4b12e8529e895/pkg/ccl/changefeedccl/poller.go#L63)):*
> ```Go
> 	details jobspb.ChangefeedDetails,
> 	highWater hlc.Timestamp,
> 	firstNextHighwater hlc.Timestamp,
> ```

oh, was this just a bug in #29611? :facepalm: The initial scan poll should be at `details.StatementTimestamp` (it needs to be selected on the gateway because we can't have each node select a different time to run the initial scan at)

so the logic here should be something like
```
if highWater == (hlc.Timestamp{}) { nextHighWater = details.StatementTime } else { nextHighWater = clock.Now }
```


<!-- Sent from Reviewable.io -->

mrtracy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29563)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/ccl/changefeedccl/poller.go, line 63 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/29563#-LMDH54c32sKa8NMsrhI:-LMDjAnl3SLyDod4Rnhe:bxqxt8m) ([raw file](https://github.com/cockroachdb/cockroach/blob/12f642e9eb643008fbd0ac7e25b4b12e8529e895/pkg/ccl/changefeedccl/poller.go#L63)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

oh, was this just a bug in #29611? :facepalm: The initial scan poll should be at `details.StatementTimestamp` (it needs to be selected on the gateway because we can't have each node select a different time to run the initial scan at)

so the logic here should be something like
```
if highWater == (hlc.Timestamp{}) { nextHighWater = details.StatementTime } else { nextHighWater = clock.Now }
```
</blockquote></details>

Done, changed the poller fix to be just this, it's still in its own commit.


<!-- Sent from Reviewable.io -->

danhhz reviewed (approved):
> Reviewed 1 of 5 files at r6, 10 of 10 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29563)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

mrtracy commented:
bors r+
